Added Settable Clock for MIMXRT BOARDS

mjs513 opened the pull request:
Per discussion on forum and discord I added a settable clock for the MIMXRT10xx boards which @KurtE have been discussing as well.  However, I used the approach of an #Ifdef in shared-bindings to test whether a settable clock is available.  I was the easiest approach for me as I could not figure out how to implement a couple of the other suggestion.  Also wanted to avoid have to update all the processor board files.  While not optimal was easiest to implement across boards.  So here goes the implementation.

1.  In each mpconfigboard.h that has a settable clock (teensy41, teensy4 and sparkfun_teensy_micromod board I added:
`#define HAS_SETTABLE_CLOCK 1`

2.  Then in sharded-bindings for processor.h in microcontroller
`uint32_t common_hal_mcu_processor_set_sys_clock(mcu_processor_obj_t *self, uint32_t frequency);
`
3.  In processor.c
```
STATIC mp_obj_t mcu_processor_set_sys_clock(mp_obj_t self_in, mp_obj_t freq) {
    mcu_processor_obj_t *self = MP_OBJ_TO_PTR(self_in);
	#if defined(HAS_SETTABLE_CLOCK)
		uint32_t value_of_freq = MP_OBJ_SMALL_INT_VALUE(freq);
		common_hal_mcu_processor_set_sys_clock(self, value_of_freq);
	#else
		mp_raise_msg(&mp_type_NotImplementedError,translate("Settable Clock Not Implemented for Your Board"));
	#endif
    return mp_const_none;
}

MP_DEFINE_CONST_FUN_OBJ_2(mcu_processor_set_sys_clock_obj, mcu_processor_set_sys_clock);
```
I added the mp_raise_msg but not sure this is the best approach.

4. In common_bindings for the mimxrt10xx boards.
processor.h I modified the struct
```
typedef struct {
    mp_obj_base_t base;
    // Stores no state currently.
	uint32_t frequency;
} mcu_processor_obj_t;
```
in processor.c
```
uint32_t common_hal_mcu_processor_set_sys_clock(mcu_processor_obj_t *self,
    uint32_t frequency) {
	SystemCoreClock = setarmclock(frequency);
    return SystemCoreClock;
}
.....
STATIC const mp_rom_map_elem_t mcu_processor_locals_dict_table[] = {
....
    { MP_ROM_QSTR(MP_QSTR_setfrequency), MP_ROM_PTR(&mcu_processor_set_sys_clock_obj) },
};
```
5. clock.h and clocks.c in ports peripherial.
```
uint32_t setarmclock(uint32_t frequency)
{
	uint32_t cbcdr = CCM->CBCDR; // pg 1021
	uint32_t cbcmr = CCM->CBCMR; // pg 1023
	uint32_t dcdc = DCDC->REG3;

	// compute required voltage
	uint32_t voltage = 1150; // default = 1.15V
	if (frequency > 528000000) {
		voltage = 1250; // 1.25V
#if defined(OVERCLOCK_STEPSIZE) && defined(OVERCLOCK_MAX_VOLT)
		if (frequency > 600000000) {
			voltage += ((frequency - 600000000) / OVERCLOCK_STEPSIZE) * 25;
			if (voltage > OVERCLOCK_MAX_VOLT) voltage = OVERCLOCK_MAX_VOLT;
		}
#endif
	} else if (frequency <= 24000000) {
		voltage = 950; // 0.95
	}

	// if voltage needs to increase, do it before switch clock speed
	CCM->CCGR6 |= CCM_CCGR6_DCDC(CCM_CCGR_ON);
	if ((dcdc & ((uint32_t)(0x1F << 0))) < ((uint32_t)(((voltage - 800) / 25) & 0x1F) << 0)) {
		//printf("Increasing voltage to %u mV\n", voltage);
........
```
The function not only changes the system clock but also changes the peripheral clock as well.

I didn't add it to .cpu.frequency intentionally btw always preferred to have it as a separate function.  This is getting long so will add some testing results next

ladyada commented:
@tannewt will likely review this when back from break. in the meantime please check actions CI to fix any formatting issues :)

dhalbert reviewed (changes requested):
I know it's easier the way you did it, but we really want to be consistent, as I mentioned in discord, the `Processor.frequency` attribute already exists as a read-only attribute. You can avoid creating a "Not Implemented" frequency setter for each and every other port by using conditional compilation in `shared-bindings` and having a default routine that throws `NotImplementedError`. For the conditional flag, add `CIRCUITPY_` to the front, something like `CIRCUITPY_SETTABLE_PROCESSOR_FREQUENCY`.  Make its default value be zero in `py/circuitpy_mpconfig.mk`, and always test it with `#if` instead of `#ifdef`. We use `#if` tests to guarantee there are no spelling errors, etc.

dhalbert commented on shared-bindings/microcontroller/Processor.c:
```diff
@@ -61,6 +72,27 @@
+STATIC mp_obj_t mcu_processor_set_sys_clock(mp_obj_t self_in, mp_obj_t freq) {
+    mcu_processor_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    #if defined(HAS_SETTABLE_CLOCK)
+    uint32_t value_of_freq = MP_OBJ_SMALL_INT_VALUE(freq);
```
This does not validate it's an int that you passed in, and that it's >= 0. Use:
```suggestion
    uint32_t value_of_freq = (uint32_t)mp_arg_validate_min(0, mp_obj_get_int(freq), MP_QSTR_frequency);
```

dhalbert replied:
Also test this by hand after you write it to make sure it's OK; I may not be completely right here.

mjs513 replied:
Done

dhalbert commented on shared-bindings/microcontroller/Processor.c:
```diff
@@ -157,6 +189,7 @@ STATIC const mp_rom_map_elem_t mcu_processor_locals_dict_table[] = {
     { MP_ROM_QSTR(MP_QSTR_temperature), MP_ROM_PTR(&mcu_processor_temperature_obj) },
     { MP_ROM_QSTR(MP_QSTR_uid), MP_ROM_PTR(&mcu_processor_uid_obj) },
     { MP_ROM_QSTR(MP_QSTR_voltage), MP_ROM_PTR(&mcu_processor_voltage_obj) },
+    { MP_ROM_QSTR(MP_QSTR_setfrequency), MP_ROM_PTR(&mcu_processor_set_sys_clock_obj) },
```
This becomes a setter on `.frequency`.
```suggestion
```

mjs513 replied:
Yep realized that as I was making the other change

dhalbert commented on shared-bindings/microcontroller/Processor.c:
```diff
@@ -61,6 +72,27 @@
+
+MP_DEFINE_CONST_FUN_OBJ_2(mcu_processor_set_sys_clock_obj, mcu_processor_set_sys_clock);
+
+const mp_obj_property_t mcu_processor_freq_obj = {
+    .base.type = &mp_type_property,
+    .proxy = {(mp_obj_t)&mcu_processor_set_sys_clock_obj,
+              MP_ROM_NONE,
+              MP_ROM_NONE},
+};
```
Restore this to the old way, and add a setter.
```suggestion
const mp_obj_property_t mcu_processor_freq_obj = {
    .base.type = &mp_type_property,
    .proxy = {(mp_obj_t)&mcu_processor_get_frequency_obj,
              (mp_obj_t)&mcu_processor_set_frequency_obj,
              MP_ROM_NONE},
};
```

mjs513 replied:
Cool didn't even realize that I did that.  

And before I forget thanks - learning something with this exercise with you walking me through things.

dhalbert replied:
Thank you for persevering!

dhalbert reviewed (changes requested):
I know it's easier the way you did it, but we really want to be consistent, as I mentioned in discord, the `Processor.frequency` attribute already exists as a read-only attribute. As you did, you can avoid creating a "Not Implemented" frequency setter for each and every other port by using conditional compilation in `shared-bindings` and having a default routine that throws `NotImplementedError`. For the conditional flag, add `CIRCUITPY_` to the front, something like `CIRCUITPY_SETTABLE_PROCESSOR_FREQUENCY`.  Make its default value be zero in `py/circuitpy_mpconfig.mk`, and always test it with `#if` instead of `#ifdef`. We use `#if` tests to guarantee there are no spelling errors, etc.

mjs513 commented on ports/mimxrt10xx/common-hal/microcontroller/Processor.c:
```diff
@@ -25,6 +25,8 @@
  * THE SOFTWARE.
  */
 
+// https://raw.githubusercontent.com/adafruit/circuitpython/main/ports/mimxrt10xx/common-hal/microcontroller/Processor.c
+
```
Not sure where that came from - I didn't add it in.  Just deleted it.

mjs513 commented on ports/mimxrt10xx/peripherals/mimxrt10xx/MIMXRT1062/clocks.c:
```diff
@@ -335,3 +336,270 @@ void clocks_init(void) {
+    }
+
+    if (!(cbcdr & CCM_CBCDR_PERIPH_CLK_SEL_L)) {
+        // printf("need to switch to alternate clock during reconfigure of ARM PLL\n");
```
Those are primarily debug comments.  I left them commented in case I wanted to go back and do some debugging if things didn't work.  Same for all the printf in clocks.c

mjs513 replied:
Just fixed all the printf.

mjs513 commented on ports/mimxrt10xx/peripherals/mimxrt10xx/MIMXRT1062/clocks.c:
```diff
@@ -335,3 +336,270 @@ void clocks_init(void) {
+            ;                                              // wait
+        }
+    } else {
+        // printf("already running from PERIPH_CLK2, safe to mess with ARM PLL\n");
```
see previous comment.

mjs513 commented on ports/mimxrt10xx/boards/teensy40/mpconfigboard.h:
```diff
@@ -16,3 +16,5 @@
 #define DEFAULT_UART_BUS_RX (&pin_GPIO_AD_B0_03)
 #define DEFAULT_UART_BUS_TX (&pin_GPIO_AD_B0_02)
+
+#define HAS_SETTABLE_CLOCK 1
```
So in mpconfigboard.mk I just add
`CIRCUITPY_SETTABLE_PROCESSOR_FREQUENCY = 1`

dhalbert replied:
And you have to add the right incantations in `py/circuitpy_mpconfig.mk` to do a `?= 0` to set a default, and to convert the make variable to a `-D...` value for the pre-processor. But there are many examples there.

mjs513 commented on shared-bindings/microcontroller/Processor.h:
```diff
@@ -39,5 +39,6 @@ mcu_reset_reason_t common_hal_mcu_processor_get_reset_reason(void);
 float common_hal_mcu_processor_get_temperature(void);
 void common_hal_mcu_processor_get_uid(uint8_t raw_id[]);
 float common_hal_mcu_processor_get_voltage(void);
+uint32_t common_hal_mcu_processor_set_sys_clock(mcu_processor_obj_t *self, uint32_t frequency);
```
Made the change locally and will push up all the changes after I validate that every still works.

mjs513 commented on shared-bindings/microcontroller/Processor.c:
```diff
@@ -61,6 +72,27 @@
+    mcu_processor_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    #if defined(HAS_SETTABLE_CLOCK)
+    uint32_t value_of_freq = MP_OBJ_SMALL_INT_VALUE(freq);
+    common_hal_mcu_processor_set_sys_clock(self, value_of_freq);
```
done

mjs513 commented on shared-bindings/microcontroller/Processor.c:
```diff
@@ -61,6 +72,27 @@
+
+STATIC mp_obj_t mcu_processor_set_sys_clock(mp_obj_t self_in, mp_obj_t freq) {
+    mcu_processor_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    #if defined(HAS_SETTABLE_CLOCK)
```
DONE

mjs513 commented on ports/mimxrt10xx/common-hal/microcontroller/Processor.c:
```diff
@@ -50,6 +52,13 @@ float common_hal_mcu_processor_get_temperature(void) {
     return temp;
 }
 
+uint32_t common_hal_mcu_processor_set_sys_clock(mcu_processor_obj_t *self,
+    uint32_t frequency) {
```
Done.

mjs513 commented:
@dhalbert 
Ok need one more lesson so I don't do this blindly (note this will also help with other things we are looking at)

In `py/circuitpy_mpconfig.mk` if we take this one line for instance
```
CIRCUITPY_USB_HOST ?= 0
CFLAGS += -DCIRCUITPY_USB_HOST=$(CIRCUITPY_USB_HOST)
```
what exactly is this doing and does that mean I have to do
```
CIRCUITPY_SETTABLE_PROCESSOR_FREQUENCY?= 0
CFLAGS += -DCIRCUITPY_SETTABLE_PROCESSOR_FREQUENCY=$(CIRCUITPY_SETTABLE_PROCESSOR_FREQUENCY)
```

to be consistent. 

Thanks
Mike

dhalbert commented:
> what exactly is this doing and does that mean I have to do
> 
> ```
> CIRCUITPY_SETTABLE_PROCESSOR_FREQUENCY?= 0
> CFLAGS += -DCIRCUITPY_SETTABLE_PROCESSOR_FREQUENCY=$(CIRCUITPY_SETTABLE_PROCESSOR_FREQUENCY)
> ```
> 
> to be consistent.

Yes, that's right. These are makefile variables. The `?= 0` means set it to that value if it is not already defined. If it's already defined earlier in the file or in an included makefile, it's not changed. Then we defined a C-preprocessor macro with the same name, by listing it on the command line: we concat it onto `CFLAGS`, which is used to build up a looong command-line argument with lots of values. You can see it if you `make` with `V=1` or `V=2`.

mjs513 commented:
Thanks Dan. Just compiled the changes for the T4.1 and ran into one issue that I have been googling on but not having much luck finding:
```
../../shared-bindings/microcontroller/Processor.c:78:40: error: implicit declaration of function 'mp_arg_validate_min'; did you mean 'mp_arg_validate_string'? [-Werror=implicit-function-declaration]
   78 |     uint32_t value_of_freq = (uint32_t)mp_arg_validate_min(24000000, mp_obj_get_int(freq), MP_QSTR_frequency);
      |                                        ^~~~~~~~~~~~~~~~~~~
      |                                        mp_arg_validate_string
../../shared-bindings/microcontroller/Processor.c:78:40: error: nested extern declaration of 'mp_arg_validate_min' [-Werror=nested-externs]
```
any clue - and where do I find definitions for things like mp_arg_validate_min?

dhalbert commented:
Oops, it's `mp_arg_validate_int_min()`, which you can find toward the end of `py/argcheck.c`, along with other validators.

The Metro M0 failed only because the build was too big; you don't need to test the board; we need to figure out how to shrink the build further.

mjs513 commented:
Thanks.  Compiling again now.

As for the M0 wanted to get SAMD board to test with as well - is there a better one to get.

dhalbert commented:
Get a SAMD51 board -- much more RAM.

mjs513 commented:
Pretty much done for a while but I did get it to compile.  However am I missing something here:
```
>>> import microcontroller
>>> microcontroller.cpu.frequency(960000000)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'int' object is not callable
```

dhalbert commented:
It is a property, so you assign to it:
`microcontroller.cpu.frequency = 960000000`

KurtE commented:
> Get a SAMD51 board -- much more RAM.
Sounds like hopefully getting close here :D 

I Also growing out a few other boards to play with these days:   ESP32-S3 ESP32-S2 
Today arrived in town two Adafruit boards: (ESP32-S2 feather and Feather STM32F05...  to try out your SDIO...)
And I have a Nano 33 IOT which I have not tried out yet... 

Will have to try this PR once it is stable...

mjs513 commented:
@dhalbert 
darnit - didnt register.

@KurtE - just ordered a micromod sam51D for testing as well.  Adafruit is out of stock on sam51d grandcentral

mjs513 commented:
@dhalbert 
think your  ` value_of_freq `  needs to read `uint32_t value_of_freq = (uint32_t)mp_arg_validate_int_min(mp_obj_get_int(freq), 24000000, MP_QSTR_frequency);` based on whats in argcheck:
but think I am going to change it `mp_arg_validate_int_range` so I can keep it to 24Mhz to 1.08Ghz

dhalbert commented:
> @dhalbert think your `value_of_freq` needs to read `uint32_t value_of_freq = (uint32_t)mp_arg_validate_int_min(mp_obj_get_int(freq), 24000000, MP_QSTR_frequency);` based on whats in argcheck: but think I am going to change it `mp_arg_validate_int_range` so I can keep it to 24Mhz to 1.08Ghz.

You may want to validate the range further in the `common_hal` routine, and raise a `ValueError` if it is out of range. The range check in `shared-bindings` is used for all potential ports, so it should only be a reasonableness check. Another port may be able to set the frequency lower or higher than 24 MHz to 1.08 GHz.

mjs513 commented:
@dhalbert - @KurtE
Think I got everything updated correctly - did do a quick test on QT PY RP2040 + T41.  I also did add the range test in common-hal just to be on the safe side.  That RP2040 takes forever to compile on my machine.

dhalbert commented:
> That RP2040 takes forever to compile on my machine.

If you are not already doing so, use `make -j<n>`, where `<n>` is the number of cores to use, to speed up the compile. See https://learn.adafruit.com/building-circuitpython/build-circuitpython#use-all-your-cpus-when-building-3086075-23

mjs513 commented:
@dhalbert @tannewt 
I made hopefully the last change.  I added a couple of messages on overclocking similar to what we do when overclocking in the Arduino ide:
```
    if (frequency <= 600000000) {
        mp_printf(&mp_plat_print, "System Clock is set to %d hz\n", frequency);
    }
    if (frequency > 600000000) {
        mp_printf(&mp_plat_print, "System Clock is set to %d hz and is Overclocked\n", frequency);
    }
    if (frequency > 816000000) {
        mp_printf(&mp_plat_print,  "System Clock is set to %d hz and is Overclocked, Cooling Required!\n", frequency);
    }
```

Thought it was appropriate.

dhalbert commented on ports/mimxrt10xx/common-hal/microcontroller/Processor.c:
```diff
@@ -57,6 +57,15 @@ uint32_t common_hal_mcu_processor_set_frequency(mcu_processor_obj_t *self,
     if (frequency < 24000000 || frequency > 1008000000) {
         mp_raise_ValueError(translate("Frequency Out of Range Must be between 24Mhz and 1.008Ghz"));
     }
+    if (frequency <= 600000000) {
+        mp_printf(&mp_plat_print, "System Clock is set to %d hz\n", frequency);
+    }
+    if (frequency > 600000000) {
+        mp_printf(&mp_plat_print, "System Clock is set to %d hz and is Overclocked\n", frequency);
+    }
+    if (frequency > 816000000) {
+        mp_printf(&mp_plat_print,  "System Clock is set to %d hz and is Overclocked, Cooling Required!\n", frequency);
```
We don't generally print out warning messages, since it can generate "noise" on stdout or the REPL. If the user is doing something like this, we expect them to read the documentation and know what they are doing. So we'd prefer that you not include these logging-style messages.

If you want to protect against overclocking, then maybe we need to add a flag like `.overclocking_allowed`, or not use the property after all to set the frequency, but instead use a function call with an extra "it's ok to overclock" flag.

mjs513 replied:
Just thought it would be a good idea but since it doesn't match your typical usage I will revert the changes.  I did see mp_warning but did notice that you don't seem to use it much.  In testing you can overclock without an issue but if running at the highest speeds like 1Ghz you should probably add cooling but as you said you can add it in the documentation somewhere.

dhalbert reviewed (changes requested):
Hi - a comment about warning messages.

mjs513 commented:
@dhalbert @KurtE @tannewt 
I changed the value error to be in line with what I saw in how micropython set up their change clock for an ESP32 (which looks rather easy to do now by the way):
```
    if (freq != 24 && freq != 150 && freq != 396 && freq != 450 && freq != 528 && freq != 600 &&
        freq != 720 && freq != 816 && freq != 912 && freq != 960 && freq != 1008) {
        mp_raise_ValueError(translate("Frequency must be 24, 150, 396, 450, 528, 600, 720, 816, 912, 960 or 1008 Mhz"));
    }
```
The frequencies match what we have been using for the T4x's and are well tested.  By way of performance testing I modified a sketch I had been using to calculate temp from clock speed using a prime number test.  It calculates 20k prime numbers per loop and then the temp is measured:
```
import time
import microcontroller
import supervisor

# Prime number calculator: https://www.educba.com/prime-number-in-c-plus-plus/
lower_init = 2
higher = 20000
supervisor.disable_autoreload()
microcontroller.cpu.frequency = 150000000
print("Sys clock set to: ", microcontroller.cpu.frequency)

while True:
    # if supervisor.runtime.serial_bytes_available:
    #    foo  # cause it to break
    lower = lower_init
    time_now = supervisor.ticks_ms()
    while (lower < higher):
        flag = 0
        for x in range(2, lower/2, 1):
            if (lower % x == 0):
                flag = 1
                break
        # if flag == 0:
        #     print("%d ", lower)
        lower = lower + 1
    print((supervisor.ticks_ms()-time_now)*0.00001667, microcontroller.cpu.temperature)

```

![CPy-clocktemp-test](https://user-images.githubusercontent.com/5366213/163727082-c0fc8114-a0f6-4617-9803-6c12272267ef.PNG)

If I remember right the temps are similar but the dt's are a lot faster in c++ implementation. Which I guess is to be expected.

mjs513 commented:
@KurtE - @dhalbert 
I ran the same (equivalent) on Micropython and got a time similar to the Arduino version of 0.006 minutes so something is off

dhalbert commented:
> @KurtE - @dhalbert I ran the same (equivalent) on Micropython and got a time similar to the Arduino version of 0.006 minutes so something is off

It may be that the frequency change is causing the time-keeping to be off. You could do some wall-clock timing to check that.

mjs513 commented:
>It may be that the frequency change is causing the time-keeping to be off. You could do some wall-clock timing to check that.

Thought crossed my mind so this morning ran the same test with the released version, 7.2.5, and the results were worse (600mhz): dt = _0.594569_ minutes. Also ran it on the qtpy rp2040 and it said 2.2 minutes for one iteration (125mhz)

mjs513 commented:
> > It may be that the frequency change is causing the time-keeping to be off. You could do some wall-clock timing to check that.
> 
> Thought crossed my mind so this morning ran the same test with the released version, 7.2.5, and the results were worse (600mhz): dt = _0.594569_ minutes. Also ran it on the qtpy rp2040 and it said 2.2 minutes for one iteration (125mhz). And wall clock seemed to be close

@KurtE - @dhalbert - @tannewt 
Not sure if this should be a whole different issue of CPy performance

mjs513 commented:
@KurtE - @dhalbert - @tannewt - @ladyada 
Ok to move completely away from the Teensy board I loaded up an Adafruit ST32F405 Feather Express board using the same Prime Number sketch in Arduino and with CPy:

Arduino Time to Calculate 20K PrimeNumbers = **0.0130 minutes**
CPy Time to Calculate 20K PrimeNumbers = **2.54334 minutes**
=== Note: S32F405 running at 168Mhz

Some other stats: 
QT PY RP2040 @125Mhz = 2.2 minutes for 20K Prime Numbers.  This is similar to the ST32F405

Teensy 4.1 @150Mhz = 1.357 minutes per 20K prime numbers.  
Teensy 4.1 @600Mhz = 0.359 minutes, Arduino = 0.003 minutes and Micropython = 0.006 minutes

Something is not quite right.  Again might be the CPy sketch itself so I really hope its something I am doing wrong.

Reference Code
Arduino Sketch:
```
long time_now;
//Prime number calculator: https://www.educba.com/prime-number-in-c-plus-plus/
int flag;
int lower_init = 2;
int higher = 20000;

void setup() {
  // put your setup code here, to run once:
  Serial.begin(115200);
  while (!Serial);
  
}

void loop() {
  int lower = lower_init;  
  time_now = millis();
  while ( lower < higher)
  {
    flag = 0;
    for ( int x = 2; x <= lower/2; ++x)
    {
      if ( lower % x == 0)
      {
        flag = 1;
        break;
      }
    }
    //if ( flag == 0){
    //  Serial.printf("%d ", lower);
    //}
    ++lower;
  }
  //Serial.println();
  Serial.println((millis()-time_now)*0.00001667,6);
}
```
CPy conversion
```
import time
import microcontroller
import supervisor

# Prime number calculator: https://www.educba.com/prime-number-in-c-plus-plus/
lower_init = 2
higher = 20000
supervisor.disable_autoreload()
# microcontroller.cpu.frequency = 150000000
print("Sys clock set to: ", microcontroller.cpu.frequency)

while True:
    # if supervisor.runtime.serial_bytes_available:
    #   foo  # cause it to break
    lower = lower_init
    time_now = supervisor.ticks_ms()
    while (lower < higher):
        flag = 0
        for x in range(2, lower/2, 1):
            if (lower % x == 0):
                flag = 1
                break
        # if flag == 0:
        #     print(lower)
        lower = lower + 1
    print((supervisor.ticks_ms()-time_now)*0.00001667, microcontroller.cpu.temperature)

```

dhalbert commented:
Please open a new issue for performance questions so it doesn't get mixed up with the original PR or issue.

This is a floating-point division. Use `/`.
`        for x in range(2, lower/2, 1):`

In the new issue, please post the MicroPython sketch as well.

mjs513 commented:
@dhalbert = @KurtE 
Will work on it

mjs513 commented:
@dhalbert @KurtE @tannewt 
Just closed the issue on CPy slowness out, https://github.com/adafruit/circuitpython/issues/6292.  Did a bit more homework and rewickered the sketch based on some speed increases when coding in MPy.  So now things make sense.  Here are some numbers for 20K prime numbers:
```
Adatfruit STM32F405 Express @168Mhz:  0.691671 minutes
ESP32S3 @240Mhz:                      0.43192 minutes
Adafruit QT PY RP2040 @135Mhz:        1.12076 minutes
Teensy 4.1 @150Mhz:                   0.58465 minutes
Teensy 4.1 @600Mhz:                   0.150 minutes
```

dhalbert commented on shared-bindings/microcontroller/Processor.c:
```diff
@@ -30,9 +30,20 @@
 #include <stdint.h>
 
-#include "py/objproperty.h"
+// #include "py/objproperty.h"
+
+// #include "py/runtime.h"
```
This looks like some kind of leftover. Just remove?

mjs513 replied:
Deleted

dhalbert commented on shared-bindings/microcontroller/Processor.c:
```diff
@@ -61,6 +72,20 @@
+    uint32_t value_of_freq = (uint32_t)mp_arg_validate_int_min(mp_obj_get_int(freq), 0, MP_QSTR_frequency);
+    common_hal_mcu_processor_set_frequency(self, value_of_freq);
+    #else
+    mp_raise_msg(&mp_type_NotImplementedError,translate("Settable Clock Not Implemented for Your Board"));
```
```suggestion
    mp_raise_msg(&mp_type_NotImplementedError,translate("frequency is read-only for this board"));
```

mjs513 replied:
Incorporated and getting ready to push

dhalbert commented on ports/mimxrt10xx/common-hal/microcontroller/Processor.c:
```diff
@@ -50,6 +52,18 @@ float common_hal_mcu_processor_get_temperature(void) {
+    uint32_t freq = frequency / 1000000;
+    if (freq != 24 && freq != 150 && freq != 396 && freq != 450 && freq != 528 && freq != 600 &&
+        freq != 720 && freq != 816 && freq != 912 && freq != 960 && freq != 1008) {
+        mp_raise_ValueError(translate("Frequency must be 24, 150, 396, 450, 528, 600, 720, 816, 912, 960 or 1008 Mhz"));
```
```suggestion
        mp_raise_ValueError(translate("Frequency must be 24, 150, 396, 450, 528, 600, 720, 816, 912, 960 or 1008 MHz"));
```

dhalbert commented on ports/mimxrt10xx/common-hal/microcontroller/Processor.c:
```diff
@@ -50,6 +52,18 @@ float common_hal_mcu_processor_get_temperature(void) {
 
+uint32_t common_hal_mcu_processor_set_frequency(mcu_processor_obj_t *self,
+    uint32_t frequency) {
+    uint32_t freq = frequency / 1000000;
+    if (freq != 24 && freq != 150 && freq != 396 && freq != 450 && freq != 528 && freq != 600 &&
+        freq != 720 && freq != 816 && freq != 912 && freq != 960 && freq != 1008) {
```
This test is not exactly for the given frequencies, since the `/ 1000000` is truncating division. E.g. 24123456 will work for 24 MHz. Is that OK with you?

mjs513 replied:
No really should be integer value otherwise the test is going to be off - will have to fix that. Guess thats another one of those // divisions

dhalbert replied:
RIght, I didn't mean it was wrong, just that it accepts inexact frequencies and truncates them to the nearest valid frequency.

dhalbert reviewed (changes requested):
OK, down to the final set of nits! Thanks for persevering. I changed the style on the error message, and have only a couple of other minor q's.

dhalbert commented:
The single CI failure is unrelated: it's an atmel-samd board whose build is too big.

mjs513 commented:
Wow - thank you now maybe will look at the stm32

tannewt reviewed (commented):
Code looks good! Thank you!

Am I reading it right that a UART at 115200 will be the correct speed before and after changing CPU frequency?

Also, why enable this on a per-board basis? Shouldn't it apply to all iMX RT?
